src/sys/system_v/shm.rs

```diff
@@ -0,0 +1,448 @@
+//! Safe wrapper around a SystemV shared memory segment
+//!
+
+use std::{
+    mem::ManuallyDrop,
+    ops::{Deref, DerefMut},
+    ptr::{null, null_mut},
+};
+
+use crate::Result;
+use crate::{errno::Errno, sys::stat::Mode};
+
+use libc::{self, c_int, c_void, key_t, shmid_ds};
+
+#[derive(Debug)]
+/// Safe wrapper around a SystemV shared memory segment
+///
+/// The shared memory segment size is equal to the size of T.
+///
+/// This is a smart pointer, and so implement the [`Deref`] and [`DerefMut`] traits.
+/// This means that you can work with the shared memory zone like you would with a [`Box`].
+///
+/// This type does not automatically create or destroy a shared memory segment,
+/// but only attach and detach from them using RAII.
+///
+/// To create one, use [`SharedMemory::shmget`], with the key [`ShmgetFlag::IPC_CREAT`].\
+/// To delete one, use [`SharedMemory::shmctl`], with the key [`ShmctlFlag::IPC_RMID`].
+///
+/// # Example
+///
+/// ```no_run
+/// # use nix::errno::Errno;
+/// # use nix::sys::system_v::shm::*;
+/// # use nix::sys::stat::Mode;
+/// #
+/// struct MyData(i64);
+/// const MY_KEY: i32 = 1337;
+///
+/// let id = SharedMemory::<MyData>::shmget(
+///     MY_KEY,
+///     ShmgetFlag::IPC_CREAT | ShmgetFlag::IPC_EXCL,
+///     Mode::S_IRWXU | Mode::S_IRWXG | Mode::S_IRWXO,
+/// )?;
+/// let mut shared_memory = SharedMemory::<MyData>::new(
+///     id,
+///     None,
+///     ShmatFlag::empty(),
+///     Mode::empty(),
+/// )?;
+///
+/// // This is writing to the stored [`MyData`] struct
+/// shared_memory.0 = 0xDEADBEEF;
+/// # Ok::<(), Errno>(())
+/// ```
+///
+pub struct SharedMemory<T> {
+    id: i32,
+    shm: ManuallyDrop<Box<T>>,
+}
+
+impl<T> Deref for SharedMemory<T> {
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target {
+        &self.shm
+    }
+}
+impl<T> DerefMut for SharedMemory<T> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.shm
+    }
+}
+
+impl<T> Drop for SharedMemory<T> {
+    fn drop(&mut self) {
+        Self::shmdt(self).expect("SharedMemory detach from SysV IPC");
+    }
+}
+
+impl<T> SharedMemory<T> {
+    /// Create a new SharedMemory object
+    ///
+    /// Attach to an existing SystemV shared memory segment.
+    ///
+    /// To create a new segment, use [`SharedMemory::shmget`], with the key [`ShmgetFlag::IPC_CREAT`].
+    ///
+    /// # Example
+    ///
+    /// ```no_run
+    /// # use nix::errno::Errno;
+    /// # use nix::sys::system_v::shm::*;
+    /// # use nix::sys::stat::Mode;
+    /// #
+    /// struct MyData(i64);
+    /// const MY_KEY: i32 = 1337;
+    ///
+    /// let mut shared_memory = SharedMemory::<MyData>::new(
+    ///     MY_KEY,
+    ///     None,
+    ///     ShmatFlag::empty(),
+    ///     Mode::empty(),
+    /// )?;
+    /// # Ok::<(), Errno>(())
+    /// ```
+    ///
+    pub fn new(
+        shmid: c_int,
+        shmaddr: Option<c_void>,
```

Comment:
It seems that one cannot create a variable of type `c_void`? (unless with `std::mem::transmute()`) Which means that the case `Some(a_value_of_type_c_void)` will never happen?

Reply:
Huh, I didn't think about that. I don't really see a way to fix it. Most use I'm finding online of `shmat` is with shmaddr set to null.
So maybe not having that feature is ok? If I remove it, I'll add it to the doc that this isn't supported.

Reply:
I hope we can have it if possible as there will always be someone who wants to use it, so if we are going to keep this argument, what is the proper type? 

1. `Option<&libc::c_void>`
  
     If `SHM_RND` is not specified in the flag argument, then the address must be page-aligned, which means in most cases, users need to manually specify a page-aligned address, and reference, is rarely page-aligned, so it is not that useful.

2. `Option<*const libc::c_void>`
    
    `Option` is rusty, but if the inner type is a raw pointer, why do we need it?

3. a simple raw pointer `*const libc::c_void`
   
   Simple, unrusty, but work.

Your thoughts on this?